broker/src/broker.rs

```diff
@@ -0,0 +1,515 @@
+        // cleanup of incompleted job, and notifying publishers when unregistering
+        // is mandatory anyway.
+        (sub_id, self.pub_txs.values().cloned().collect(), announce)
+    }
+
+    // Unregister the subscriber. Similar to register_subscriber, returns list
+    // of channels through which publishers must be notified about the removal.
+    pub fn unregister_subscriber(
+        &mut self,
+        sub_id: SubId,
+        sub_key: SubscriptionKey,
+    ) -> Option<(Vec<Sender<SubAnnounce>>, SubAnnounce)> {
+        // We need to notify existing publishers only about per timeline
+        // subscriptions, 'all' kind is detected on its own through closed
+        // channels.
+        let announce = match sub_key {
+            SubscriptionKey::All => {
+                assert!(self.subs_to_all.contains_key(&sub_id));
+                self.subs_to_all.remove(&sub_id);
+                None
+            }
+            SubscriptionKey::Timeline(ref ttid) => {
+                remove_sub(&mut self.subs_to_timelines, ttid, sub_id);
+                Some(SubAnnounce::RemoveTimeline(*ttid, sub_id))
+            }
+        };
+        announce.map(|a| (self.pub_txs.values().cloned().collect(), a))
+    }
+
+    pub fn report(&mut self) {
+        println!(
+            "registered {} publishers, {} subs to all, {} subs to timelines",
+            self.pub_txs.len(),
+            self.subs_to_all.len(),
+            self.subs_to_timelines.len(),
+        );
+    }
+}
+
+// SharedState wrapper for post-locking operations (sending to pub_tx chans).
+#[derive(Clone)]
+struct Registry {
+    shared_state: Arc<Mutex<SharedState>>,
+}
+
+impl Registry {
+    // Register new publisher in shared state.
+    pub fn register_publisher(&self) -> Publisher {
+        let (announce_tx, announce_rx) = mpsc::channel(128);
+        let mut ss = self.shared_state.lock().unwrap();
+        let id = ss.register_publisher(announce_tx);
+        let (subs_to_all, subs_to_timelines) = (
+            ss.subs_to_all.values().cloned().collect(),
+            ss.subs_to_timelines.clone(),
+        );
+        drop(ss);
+        // println!("registered publisher {}", id);
+        Publisher {
+            id,
+            announce_rx: announce_rx.into(),
+            subs_to_all,
+            subs_to_timelines,
+            registry: self.clone(),
+        }
+    }
+
+    pub fn unregister_publisher(&self, publisher: &Publisher) {
+        self.shared_state
+            .lock()
+            .unwrap()
+            .unregister_publisher(publisher.id);
+        // println!("unregistered publisher {}", publisher.id);
+    }
+
+    // Register new subscriber in shared state.
+    pub async fn register_subscriber(&self, sub_key: SubscriptionKey) -> Subscriber {
+        let (tx, rx) = mpsc::channel(CHAN_SIZE);
+        let id;
+        let mut pub_txs;
+        let announce;
+        {
+            let mut ss = self.shared_state.lock().unwrap();
```

Comment:
Does it actually make any sense to use `Mutex` from `std`? I see plenty of async code around it, feels like the `tokio` one could be more suitable?

Reply:
There is no await inside, so std is fine.